internal/k8s/controllers/config_controller_test.go

```diff
@@ -0,0 +1,335 @@
+						Name:      "adv2",
+						Namespace: testNamespace,
+					},
+					Spec: v1beta1.BGPAdvertisementSpec{
+						AggregationLength:   pointer.Int32Ptr(24),
+						AggregationLengthV6: pointer.Int32Ptr(64),
+						IPAddressPools:      []string{"pool1"},
+					},
+				},
+				{
+					ObjectMeta: v1.ObjectMeta{
+						Name:      "adv3",
+						Namespace: testNamespace,
+					},
+					Spec: v1beta1.BGPAdvertisementSpec{
+						IPAddressPools: []string{"pool2"},
+					},
+				},
+			},
+			L2Advs: []v1beta1.L2Advertisement{
+				{
+					ObjectMeta: v1.ObjectMeta{
+						Name:      "l2adv1",
+						Namespace: testNamespace,
+					},
+					Spec: v1beta1.L2AdvertisementSpec{
+						IPAddressPools: []string{"pool3"},
+					},
+				},
+				{
+					ObjectMeta: v1.ObjectMeta{
+						Name:      "l2adv2",
+						Namespace: testNamespace,
+					},
+				},
+			},
+			PasswordSecrets: map[string]corev1.Secret{
+				"bgpsecret": {Type: corev1.SecretTypeBasicAuth, ObjectMeta: v1.ObjectMeta{Name: "bgpsecret", Namespace: testNamespace},
+					Data: map[string][]byte{"password": []byte([]byte("nopass"))}},
+			},
+			LegacyAddressPools: []v1beta1.AddressPool{
+				{
+					ObjectMeta: v1.ObjectMeta{
+						Name:      "legacypool1",
+						Namespace: testNamespace,
+					},
+					Spec: v1beta1.AddressPoolSpec{
+						Addresses: []string{
+							"10.21.0.0/16",
+							"10.51.0.0/24",
+						},
+						Protocol:      "bgp",
+						AvoidBuggyIPs: true,
+						AutoAssign:    pointer.BoolPtr(false),
+					},
+				},
+			},
+		}
+
+		calledForceReload = false
+		mockForceReload   = func() { calledForceReload = true }
+	)
+
+	if err := v1beta1.AddToScheme(scheme); err != nil {
+		t.Fatalf("add to scheme failed: %s", err)
+	}
+
+	if err := v1beta2.AddToScheme(scheme); err != nil {
+		t.Fatalf("add to scheme failed: %s", err)
+	}
+
+	if err := corev1.AddToScheme(scheme); err != nil {
+		t.Fatalf("add to scheme failed: %s", err)
+	}
+
+	tests := []struct {
+		desc         string
+		retSyncState SyncState
+	}{
+		{
+			desc:         "handler returns SyncStateSuccess",
+			retSyncState: SyncStateSuccess,
+		},
+		{
+			desc:         "handler returns SyncStateError",
+			retSyncState: SyncStateError,
+		},
+		{
+			desc:         "handler returns SyncStateErrorNoRetry",
+			retSyncState: SyncStateErrorNoRetry,
+		},
+		{
+			desc:         "handler returns SyncStateReprocessAll",
+			retSyncState: SyncStateReprocessAll,
+		},
+	}
+	for _, test := range tests {
+		calledForceReload = false
+		expectedCfg, err := config.For(resources, config.DontValidate)
+		if err != nil {
+			t.Fatalf("test %s failed, got unexpected error: %v", test.desc, err)
+		}
+
+		initObjects := objectsFromResources(resources)
+		fakeClient := fake.NewClientBuilder().WithScheme(scheme).WithObjects(initObjects...).Build()
+
+		mockHandler := func(l log.Logger, cfg *config.Config) SyncState {
+			if diff := cmp.Diff(expectedCfg, cfg, cmpopts.IgnoreUnexported(metallbcfg.Pool{})); diff != "" {
```

Comment:
use cmp.Cmp and print the diff in case of failure

Reply:
done!